### PR TITLE
Show clearer error message when #![deny(warnings)] escalates a warning

### PR DESCRIPTION
**Pull request by shssoichiro**
Addresses #30730

### CODE REVIEW

**Comment by oli-obk** [comment]
This can probably use a compile-fail test

**Comment by shssoichiro** [comment]
Added, hopefully that's complete enough. I'm not familiar writing tests for rustc, but it passed when I ran `make check`.

File: src/test/compile-fail/issue-30730.rs [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![deny(warnings)]
```
Comment (oli-obk) [inline src/test/compile-fail/issue-30730.rs]: this should also have a comment saying `//~ NOTE: lint level defined here`

**Comment by oli-obk** [comment]
`make check-stage1 TESTNAME=blub` (you can find this under `make tips`)

File: src/test/compile-fail/issue-30730.rs (hunk 1 above)
Comment (shssoichiro) [inline src/test/compile-fail/issue-30730.rs]: Okay, that's what I was unsure of. I've pushed another commit adding this.

**Comment by oli-obk** [comment]
ping @Aatch

**Comment by nrc** [comment]
Apologies for the late review, I've been on vacation.

File: src/librustc/lint/context.rs [hunk 2]
```diff
@@ -451,9 +451,14 @@ pub trait LintContext: Sized {
 
     fn level_src(&self, lint: &'static Lint) -> Option<LevelSource> {
         self.lints().levels.get(&LintId::of(lint)).map(|ls| match ls {
-            &(Warn, src) => {
+            &(Warn, _) => {
                 let lint_id = LintId::of(builtin::WARNINGS);
-                (self.lints().get_level_source(lint_id).0, src)
+                let warn_src = self.lints().get_level_source(lint_id);
+                if warn_src.0 != Warn {
+                    warn_src
```
Comment (nrc) [inline src/librustc/lint/context.rs]: maybe better to make warn_src mutable

Comment (nrc) [inline src/librustc/lint/context.rs]: or not, actually. Sorry.

**Pull request merged**